subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+		LOG_WRN("File system can be mounted read only. RO features %d detected.",
+				(sb->s_feature_ro_compat & ~EXT2_FEATURE_RO_COMPAT_SUPPORTED));
+		return -EROFS;
+	}
+
+	LOG_DBG("ino_cnt:%d blk_cnt:%d blk_per_grp:%d ino_per_grp:%d free_ino:%d free_blk:%d "
+			"blk_size:%d ino_size:%d mntc:%d",
+			sb->s_inodes_count, sb->s_blocks_count, sb->s_blocks_per_group,
+			sb->s_inodes_per_group, sb->s_free_inodes_count, sb->s_free_blocks_count,
+			1024 << sb->s_log_block_size, sb->s_inode_size, sb->s_mnt_count);
+	return 0;
+}
+
+int ext2_init_fs(struct ext2_data *fs)
+{
+	int ret = 0;
+
+	/* Fetch superblock */
+	if (fs->block_size == 1024) {
+		fs->sblock_offset = 0;
+		fs->sblock = ext2_get_block(fs, 1);
+	} else {
+		fs->sblock_offset = 1024;
+		fs->sblock = ext2_get_block(fs, 0);
+	}
+
+	if (fs->sblock == NULL) {
+		ret = ENOENT;
+		goto out;
+	}
+
+	if (!(fs->flags & EXT2_DATA_FLAGS_RO)) {
+		/* Update sblock fields set during the successful mount. */
+		EXT2_DATA_SBLOCK(fs)->s_state = EXT2_ERROR_FS;
+		EXT2_DATA_SBLOCK(fs)->s_mnt_count += 1;
+		ret = ext2_write_block(fs, fs->sblock);
+		if (ret < 0) {
+			goto out;
+		}
+	}
+
+	ret = ext2_fetch_block_group(fs, 0);
+	if (ret < 0) {
+		goto out;
+	}
+
+	ret = ext2_fetch_bg_ibitmap(fs->bgroup);
+	ret = ext2_fetch_bg_bbitmap(fs->bgroup);
+
+	/* Validate superblock */
+	uint32_t set;
+	struct ext2_disk_superblock *sb = EXT2_DATA_SBLOCK(fs);
+	uint32_t fs_blocks = sb->s_blocks_count - sb->s_first_data_block;
+
+	set = ext2_bitmap_count_set(BGROUP_BLOCK_BITMAP(fs->bgroup), fs_blocks);
+
+	if (set != sb->s_blocks_count - sb->s_free_blocks_count - sb->s_first_data_block) {
+		error_beharior(fs, "Wrong number of used blocks in superblock and bitmap");
+		return -EINVAL;
+	}
+
+	set = ext2_bitmap_count_set(BGROUP_INODE_BITMAP(fs->bgroup), sb->s_inodes_count);
+
+	if (set != sb->s_inodes_count - sb->s_free_inodes_count) {
+		error_beharior(fs, "Wrong number of used inodes in superblock and bitmap");
+		return -EINVAL;
+	}
+	return 0;
+out:
+	ext2_drop_block(fs->sblock);
+	fs->sblock = NULL;
+	return ret;
+}
+
+int ext2_close_fs(struct ext2_data *fs)
+{
+	int ret = 0;
+
+	/* Close all open inodes */
+	for (int32_t i = 0; i < fs->open_inodes; ++i) {
+		if (fs->inode_pool[i] != NULL) {
+			ext2_inode_drop(fs->inode_pool[i]);
+		}
+	}
+
+	if (!(fs->flags & EXT2_DATA_FLAGS_RO)) {
+		EXT2_DATA_SBLOCK(fs)->s_state = EXT2_VALID_FS;
+		ret = ext2_write_block(fs, fs->sblock);
+		if (ret < 0) {
+			return -EIO;
+		}
+	}
+
+	/* free block group if it is fetched */
+	if (fs->bgroup != NULL) {
+		ext2_drop_block(fs->bgroup->inode_table);
+		ext2_drop_block(fs->bgroup->inode_bitmap);
+		ext2_drop_block(fs->bgroup->block_bitmap);
+		ext2_drop_block(fs->bgroup->block);
+	}
+
+	if (fs->sync(fs) < 0) {
+		return -EIO;
+	}
+	ext2_drop_block(fs->sblock);
+	fs->sblock = NULL;
+	return 0;
+}
+
+int ext2_close_struct(struct ext2_data *fs)
+{
+	memset(fs, 0, sizeof(struct ext2_data));
```

Comment:
This should probably be `sizeof(*fs)`, rather than sizeof on type directly.

Reply:
Already I think that it looks more clear when we take sizeof the type rather than variable. We immediately see what size we need (because we know which type definition to look up).

Although, if it is the convention in Zephyr I will change that.